Framework/PythonInterface/mantid/api/src/Exports/IFunction.cpp

```diff
@@ -35,6 +36,32 @@ PyObject *getCategories(IFunction &self) {
   return registered;
 }
 
+/// An object for constructing a shared_ptr that won't ever delete its pointee
+class NoDeleting {
+public:
+  /// Does nothing
+  void operator()(void *) {}
+  /// Does nothing
+  void operator()(const void *) {}
+};
+
+//------------------------------------------------------------------------------------------------------
+/**
+* Something that makes IFunction return to python a composite function
+* for Product function, Convolution or
+* any similar superclass of composite function.
+* @param self :: Enables it to be called as a member function on the
+* IFunction class
+*/
+Mantid::API::CompositeFunction_sptr castToCompositeFunction(IFunction &self) {
```

Comment:
The ``+`` operator has no such issue, because unlike ProductFunction, CompositeFunction receives its component composite functions as CompositeFunction. It is this peculiarity of the python binding that is the issue.